x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -929,4 +969,66 @@ private boolean verify(SecureString password) {
             return hash != null && cacheHasher.verify(password, hash);
         }
     }
+
+    public static final class ApiKeyDoc {
+
+        static final InstantiatingObjectParser<ApiKeyDoc, Void> PARSER;
+        static {
+            InstantiatingObjectParser.Builder<ApiKeyDoc, Void> builder =
+                InstantiatingObjectParser.builder("api_key_doc", true, ApiKeyDoc.class);
+            builder.declareString(constructorArg(), new ParseField("doc_type"));
+            builder.declareLong(constructorArg(), new ParseField("creation_time"));
+            builder.declareLongOrNull(constructorArg(), -1, new ParseField("expiration_time"));
+            builder.declareBoolean(constructorArg(), new ParseField("api_key_invalidated"));
+            builder.declareString(optionalConstructorArg(), new ParseField("api_key_hash"));
+            builder.declareString(constructorArg(), new ParseField("name"));
+            builder.declareInt(constructorArg(), new ParseField("version"));
+            ObjectParserHelper<ApiKeyDoc, Void> parserHelper = new ObjectParserHelper<>();
+            parserHelper.declareRawObject(builder, optionalConstructorArg(), new ParseField("role_descriptors"));
+            parserHelper.declareRawObject(builder, constructorArg(), new ParseField("limited_by_role_descriptors"));
+            parserHelper.declareRawObject(builder, constructorArg(), new ParseField("creator"));
+            PARSER = builder.build();
```

Comment:
Is it possible that some fields, e.g. `version` or `name`, are completely missing from the source document? This parser requires all fields to present (`null` value is ok but it must present). In current code, all fields do present. I wonder whether this is true for older versions.

Reply:
I suspect it's not true for all versions, or at least it might not be.
We'd probably want to generate an API key from v6.7 and check, but it might be safer to change the parser to be more accommodating of missing fields.

Reply:
I checked v6.7 and it looks fine, i.e. it has the same fields as in the master branch. But it might still be safer to have more leniency for missing fields? I'll see how it can be done.

Reply:
It's actually fairly easily to make fields optional by using 

`builder.declareString(ApiKeyDoc::setName, new ParseField("name"))` 

instead of 

`builder.declareString(constructorArg(), new ParseField("name"))`.

But I ended up didn't make any of the fields optional because:
* All of them exist from 6.7. So not sure which one I should pick for optional. (The only thing missing in 6.7 is `creator.realm_type`. But it is still wrapped inside a `Map<String, Object>` in the new implementation so it is not an issue).
* Once a field made optional, it cannot be `final` and becomes an anomaly in current implementation. It's not a strong argument, just a personal preference.

Please let me know if you have concerns.